File: iosclient/ReadingApp/ReadingApp/Utilities/StudyDurationTracker.swift
import Foundation

/// 前台有效学习时长追踪：活跃时累计秒数，定期/退出时通过回调上报。
@MainActor
final class StudyDurationTracker {
    private let idleTimeout: TimeInterval
    private let flushInterval: TimeInterval
    private let report: (Int) async -> Void

    private var accumulatedSeconds = 0
    private var segmentStartedAt: Date?
    private var lastInteractionAt = Date()
    private var isTracking = false
    private var isAppActive = true
    private var flushTask: Task<Void, Never>?

    init(
        idleTimeout: TimeInterval = 60,
        flushInterval: TimeInterval = 15,
        report: @escaping (Int) async -> Void
    ) {
        self.idleTimeout = idleTimeout
        self.flushInterval = flushInterval
        self.report = report
    }

    func start() {
        guard !isTracking else { return }
        isTracking = true
        lastInteractionAt = Date()
        if isAppActive {
            beginSegment()
        }
        startFlushLoop()
    }

    func stop() {
        guard isTracking else { return }
        isTracking = false
        flushTask?.cancel()
        flushTask = nil
        endSegment()
        Task {
            await flushNow()
        }
    }

    func noteInteraction() {
        lastInteractionAt = Date()
        guard isTracking, isAppActive else { return }
        if segmentStartedAt == nil {
            beginSegment()
        }
    }

    func setAppActive(_ active: Bool) {
        isAppActive = active
        guard isTracking else { return }
        if active {
            lastInteractionAt = Date()
            beginSegment()
        } else {
            endSegment()
            Task { await flushNow() }
        }
    }

    private func beginSegment() {
        if segmentStartedAt == nil {
            segmentStartedAt = Date()
        }
    }

    private func endSegment() {
        guard let started = segmentStartedAt else { return }
        let now = Date()
        let idleCut = lastInteractionAt.addingTimeInterval(idleTimeout)
        let effectiveEnd = min(now, idleCut)
        let seconds = max(0, Int(effectiveEnd.timeIntervalSince(started)))
        if seconds > 0 {
            accumulatedSeconds += seconds
        }
        segmentStartedAt = nil
    }

    private func startFlushLoop() {
        flushTask?.cancel()
        flushTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64((self?.flushInterval ?? 15) * 1_000_000_000))
                guard !Task.isCancelled else { return }
                await self?.tickAndFlush()
            }
        }
    }

    private func tickAndFlush() async {
        guard isTracking else { return }
        if isAppActive {
            let idleDeadline = lastInteractionAt.addingTimeInterval(idleTimeout)
            if Date() > idleDeadline {
                endSegment()
            } else if segmentStartedAt == nil {
                beginSegment()
            } else {
                // 周期性截断当前片段，避免长时间会话只在退出时上报。
                endSegment()
                beginSegment()
            }
        }
        await flushNow()
    }

    private func flushNow() async {
        let seconds = accumulatedSeconds
        guard seconds > 0 else { return }
        accumulatedSeconds = 0
        await report(seconds)
    }
}

File: console/web/src/utils/studyDurationTracker.ts
type ReportFn = (seconds: number) => Promise<void>;

/** 浏览器端有效学习时长：前台可见且未空闲时累计，定期上报。 */
export class StudyDurationTracker {
  private accumulated = 0;
  private segmentStartedAt: number | null = null;
  private lastInteractionAt = Date.now();
  private tracking = false;
  private visible = typeof document === "undefined" ? true : !document.hidden;
  private timer: number | null = null;
  private readonly idleTimeoutMs: number;
  private readonly flushIntervalMs: number;
  private readonly report: ReportFn;

  constructor(
    report: ReportFn,
    options?: { idleTimeoutMs?: number; flushIntervalMs?: number },
  ) {
    this.report = report;
    this.idleTimeoutMs = options?.idleTimeoutMs ?? 60_000;
    this.flushIntervalMs = options?.flushIntervalMs ?? 15_000;
  }

  start() {
    if (this.tracking) return;
    this.tracking = true;
    this.lastInteractionAt = Date.now();
    this.visible = !document.hidden;
    if (this.visible) this.beginSegment();
    this.timer = window.setInterval(() => {
      void this.tickAndFlush();
    }, this.flushIntervalMs);
    document.addEventListener("visibilitychange", this.onVisibility);
    window.addEventListener("pointerdown", this.onInteraction);
    window.addEventListener("keydown", this.onInteraction);
  }

  stop() {
    if (!this.tracking) return;
    this.tracking = false;
    if (this.timer != null) {
      window.clearInterval(this.timer);
      this.timer = null;
    }
    document.removeEventListener("visibilitychange", this.onVisibility);
    window.removeEventListener("pointerdown", this.onInteraction);
    window.removeEventListener("keydown", this.onInteraction);
    this.endSegment();
    void this.flushNow();
  }

  noteInteraction() {
    this.lastInteractionAt = Date.now();
    if (this.tracking && this.visible && this.segmentStartedAt == null) {
      this.beginSegment();
    }
  }

  private onVisibility = () => {
    this.visible = !document.hidden;
    if (!this.tracking) return;
    if (this.visible) {
      this.lastInteractionAt = Date.now();
      this.beginSegment();
    } else {
      this.endSegment();
      void this.flushNow();
    }
  };

  private onInteraction = () => {
    this.noteInteraction();
  };

  private beginSegment() {
    if (this.segmentStartedAt == null) {
      this.segmentStartedAt = Date.now();
    }
  }

  private endSegment() {
    if (this.segmentStartedAt == null) return;
    const now = Date.now();
    const idleCut = this.lastInteractionAt + this.idleTimeoutMs;
    const effectiveEnd = Math.min(now, idleCut);
    const seconds = Math.max(
      0,
      Math.floor((effectiveEnd - this.segmentStartedAt) / 1000),
    );
    if (seconds > 0) this.accumulated += seconds;
    this.segmentStartedAt = null;
  }

  private async tickAndFlush() {
    if (!this.tracking) return;
    if (this.visible) {
      if (Date.now() > this.lastInteractionAt + this.idleTimeoutMs) {
        this.endSegment();
      } else if (this.segmentStartedAt == null) {
        this.beginSegment();
      } else {
        this.endSegment();
        this.beginSegment();
      }
    }
    await this.flushNow();
  }

  private async flushNow() {
    const seconds = this.accumulated;
    if (seconds <= 0) return;
    this.accumulated = 0;
    try {
      await this.report(seconds);
    } catch {
      this.accumulated += seconds;
    }
  }
}
